samples/isprime/isprime.neon

```diff
@@ -0,0 +1,211 @@
+%| Rabin/Miller test to test a prime number.                   
+   n - 1 = 2km for some k > 0 and odd integer m.               
+   a = 1 < a < n - 1                                           
+   We say that n passes the Rabin-Miller test for the base a if 
+   either the first term in this sequence is 1, or 1 occurs later 
+   in this sequence and is immediately preceded by -1. If the test 
+   fails, we know that n is composite, and describe the base a as 
+   a witness to the compositeness of n. The crucial extra information 
+   we can derive is that if (odd) n is composite, then at least 3/4 
+   of the numbers a with 1 < a < n - 1 are witness to this fact.
+   First calculate m, in Neon we do this by simply shifting the bits
+   right k times, until m is odd and less than n.
+   Next, we calculate a, such that a is less than n - 1, and a > 1.
+   Then, we calculate a ^ m mod n.  If this result is 1, then
+   we need to generate a new "a" and try again.
+ |%
+IMPORT binary
+IMPORT random
+IMPORT string
+IMPORT sys
+
+EXPORT RabinMiller
+EXPORT SmallFactor
+
+LET SMALL_PRIMES: Array<Number> := [  2,   3,   5,   7,  11,  13,  17,  19,  23,  29,  31,  37,  41,  43,  47,  53,  59,  61, 
+                                     67,  71,  73,  79,  83,  89,  97, 101, 103, 107, 109, 113, 127, 131, 137, 139, 149, 151, 
+                                    157, 163, 167, 173, 179, 181, 191, 193, 197, 199, 211, 223, 227, 229, 233, 239, 241, 251, 257 ]
+LET iDefaultTests:  Number      := 5
+VAR TestIterations: Number      := iDefaultTests % k in the algorithm.  This can be adjusted by the parameter -k.
+VAR bQuiet:         Boolean     := FALSE
+
+
+FUNCTION GetRandomA(n:Number): Number
+    RETURN (random.uint32() MOD (n - 2)) + 1
+END FUNCTION
+
+FUNCTION isOdd(n: Number): Boolean
+    RETURN binary.get32(n, 0)
+END FUNCTION
+
+FUNCTION RabinMiller(n: Number, iIterations: Number): Boolean
+    VAR m:       Number := n - 1
+    VAR k:       Number := 0
+    VAR a:       Number := 0
+
+    IF NOT isOdd(n) THEN
+        % Rule: The number n is an EVEN number, and therefore CANNOT be prime.
+        IF n = 2 THEN
+            % The number 2 is the only exception to the rule!
+            % Since 2 is the ONLY even number that is prime, we handle this
+            % case a little differently.
+            RETURN TRUE
+        END IF
+        RETURN FALSE
+    END IF
+
+    IF m > 0 THEN
+        WHILE binary.get32(m, 0) DO
+            INC k
+            m := binary.shiftRight32(m, 1)
+        END WHILE
+    END IF
+
+    VAR r: Number
+    FOR iCount := 0 TO iIterations LABEL WitnessLoop DO
+        a := GetRandomA(n)
+        r := 1
+        FOR x := 0 TO m - 1 DO
```

Comment:
If you just want to do something `m` times, it's probably clearer to write `FOR x := 1 TO m`. Saves a subtraction too.